stdlib/public/Concurrency/TaskPrivate.h

```diff
@@ -206,31 +268,99 @@ class alignas(sizeof(void*) * 2) ActiveTaskStatus {
   ActiveTaskStatus() = default;
 #endif
 
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+  constexpr ActiveTaskStatus(JobPriority priority)
+      : Flags(uintptr_t(priority)), DrainLock(DLOCK_OWNER_NULL), Record(nullptr) {}
+#else
   constexpr ActiveTaskStatus(JobPriority priority)
-      : Record(nullptr), Flags(uintptr_t(priority)) {}
+      : Flags(uintptr_t(priority)), Record(nullptr) {}
+#endif
 
   /// Is the task currently cancelled?
   bool isCancelled() const { return Flags & IsCancelled; }
   ActiveTaskStatus withCancelled() const {
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+    return ActiveTaskStatus(Record, Flags | IsCancelled, DrainLock);
+#else
     return ActiveTaskStatus(Record, Flags | IsCancelled);
+#endif
   }
 
-  /// Is the task currently running?
+  bool isEnqueued() const { return Flags & IsEnqueued; }
+  ActiveTaskStatus withEnqueued() const {
```

Comment:
I don't know where the right funnel point in code for when a task is enqueued - whether it is on an actor or a global pool. If we had such a funnel point, it would make it easier to

(a) Set the enqueued bit on the task state
(b) Copy over the maxPriority from active task status to the job field. Right now, this operation happens in a bunch of places and could be consolidated to one funnel point.

All the entry points that I have found for enqueuing work onto an executor only take Job structures though. Should I provide overloads for all of the `swift_task_enqueue` functions such that they task Task objects?

Reply:
I think you'll want to add a method on `Job` that you call from the enqueue functions. Then that method can do a dynamic cast and call a Task method, like, `if (auto task = dyn_cast<AsyncTask>(this)) task->handleEnqueueing()`, to do task-specific stuff.

Reply:
Where do we actually want to do the record-keeping associated with enqueueing a task?  My intuition is that in most cases we'll want to take care of this in the caller of `swift_task_enqueue` because we'll have to say that the task is no longer waiting on whatever it was waiting on, but maybe that's not true and it should be part of `swift_task_enqueue`.

Reply:
Let me take a stab at doing it in each caller and seeing how that looks. I think you might be right that we might want to do something different based on the different calling contexts to `swift_task_enqueue` and its friends.

Since we have multiple flavors of enqueue functions as well, I was hoping to avoid having to hunt down each caller.

Reply:
I've taken a stab at fixing this and I think this actually really cleans up the state machine of identifying the various states a task can be in. Feedback appreciated!